poc/four/main.py

```diff
@@ -0,0 +1,267 @@
+        else:
+            # Initialize OAuth2 client
+            oauth_client = OAuth2Client(
+                oauth_endpoints['token'],
+                config['client_id'],
+                config['client_secret']
+            )
+            logger.info("OAuth2 client initialized")
+            
+            # Initialize Ed-Fi API client
+            api_client = EdFiApiClient(data_api_base, oauth_client)
+            logger.info("Ed-Fi API client initialized")
+        
+        # Load and validate CSV data
+        data_handler = StudentDataHandler(csv_file)
+        df = data_handler.load_data()
+        
+        if not data_handler.validate_data(df):
+            logger.error("Data validation failed")
+            return False
+        
+        # Group records by action
+        records_by_action = data_handler.get_records_by_action(df)
+        
+        # Process each action type
+        success_count = 0
+        error_count = 0
+        
+        for action, records in records_by_action.items():
+            if not records:
+                continue
+                
+            logger.info(f"Processing {len(records)} {action} operations")
+            
+            for record in records:
+                try:
+                    if dry_run:
+                        logger.info(f"[DRY RUN] Would {action}: {record}")
+                        success_count += 1
+                        continue
+                    
+                    success = False
+                    
+                    if action == 'insert':
+                        # Create student
+                        if api_client.create_student(record['student']):
+                            success = True
+                            # Create enrollment if data exists
+                            if record['enrollment']:
+                                if not api_client.create_enrollment(record['enrollment']):
+                                    logger.warning(f"Student created but enrollment failed for {record['student']['studentUniqueId']}")
+                        
+                    elif action == 'update':
+                        # Update student
+                        student_id = record['student']['studentUniqueId']
+                        if api_client.update_student(student_id, record['student']):
+                            success = True
+                            # Update enrollment if data exists
+                            if record['enrollment']:
+                                school_id = record['enrollment']['schoolReference']['schoolId']
+                                if not api_client.update_enrollment(student_id, school_id, record['enrollment']):
+                                    logger.warning(f"Student updated but enrollment update failed for {student_id}")
+                        
+                    elif action == 'delete':
+                        # Delete student (this should cascade to enrollments)
```

Comment:
That's a bold assertion - and it is wrong. This will leave an orphaned `studentSchoolAssociation` record. Need to first delete the `student` then delete the `studentSchoolAssociation`. Another good point to make in a future prompt, which will need to include additional business logic.